0000-clamp.md

```diff
@@ -0,0 +1,155 @@
+- Feature Name: clamp functions
+- Start Date: 2017-03-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add functions to the language which take a value and an inclusive range, and will "clamp" the input to the range.  I.E.
+
+```Rust
+if input > max {
+    return max;
+}
+else if input < min {
+    return min;
+} else {
+    return input;
+}
+```
+
+Likely locations would be in std::cmp::clamp implemented for all Ord types, and a special version implemented for f32 and f64.
+The f32 and f64 versions could live either in std::cmp or in the primitive types themselves.  There are good arguments for either
+location.
+
+# Motivation
+[motivation]: #motivation
+
+Clamp is a very common pattern in Rust libraries downstream.  Some observed implementations of this include:
+
+http://nalgebra.org/rustdoc/nalgebra/fn.clamp.html
+
+http://rust-num.github.io/num/num/fn.clamp.html
+
+Many libraries don't expose or consume a clamp function but will instead use patterns like this:
+```Rust
+if input > max {
+    max
+}
+else if input < min {
+    min
+} else {
+    input
+}
+```
+and
+```Rust
+input.max(min).min(max);
+```
+and even
+```Rust
+match input {
+    c if c >  max =>  max,
+    c if c < min => min,
+    c              =>  c,
+}
+```
+
+Typically these patterns exist where there is a need to interface with APIs that take normalized values or when sending 
+output to hardware that expects values to be in a certain range, such as audio samples or painting to pixels on a display.
+
+While this is pretty trivial to implement downstream there are quite a few ways to do it and just writing the clamp 
+inline usually results in rather a lot of control flow structure to describe a fairly simple and common concept.
+
+# Detailed design
+[design]: #detailed-design
+
+Add the following to std::cmp
+
+```Rust
+use ops::RangeInclusive;
+/// Returns the upper bound of the range if input is greater than the range, and the lower bound of
+/// the range if input is less than the range.  Otherwise this will return input.
+#[inline]
+pub fn clamp<T: Ord>(input: T, range: RangeInclusive<T>) -> T {
```

Comment:
I'm not sure if we should tie the stability of this RFC with #1192 (rust-lang/rust#28237) here by using `range: RangeInclusive<T>` instead of `min: T, max: T`.